### PR TITLE
[Refactor] Fixing mypy issues for `syft/service/` and `syft/util/`

### PR DESCRIPTION
**Pull request by khoaguin**
Fix mypy issues for `syft/service` and `syft/util` directories

### CODE REVIEW

File: packages/syft/tests/syft/request/request_multiple_nodes_test.py [hunk 1]
```diff
@@ -149,6 +150,7 @@ def compute_sum(data) -> float:
 
 
 @pytest.mark.flaky(reruns=2, reruns_delay=1)
+@pytest.mark.skipif(sys.platform == "win32", reason="very flaky on windows")
```
Comment (kiendang) [inline packages/syft/tests/syft/request/request_multiple_nodes_test.py]: This should be in a separate PR. You can create a separate PR for this and post it in the engineering channel to have others' opinion.

File: packages/syft/src/syft/service/action/action_graph.py [hunk 2]
```diff
@@ -79,7 +78,7 @@ def make_created_at(cls, v: Optional[DateTime]) -> DateTime:
         return DateTime.now() if v is None else v
 
     @staticmethod
-    def from_action(action: Action, credentials: SyftVerifyKey):
+    def from_action(action: Action, credentials: SyftVerifyKey) -> "NodeActionData":
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_graph.py]: This should be a `classmethod` returning `Self`. `return cls(...` instead of `return NodeActionData(...`.

File: packages/syft/src/syft/service/action/action_graph.py [hunk 3]
```diff
@@ -91,24 +90,26 @@ def from_action(action: Action, credentials: SyftVerifyKey):
         )
 
     @staticmethod
-    def from_action_obj(action_obj: ActionObject, credentials: SyftVerifyKey):
+    def from_action_obj(
+        action_obj: ActionObject, credentials: SyftVerifyKey
+    ) -> "NodeActionData":
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_graph.py]: same here

File: packages/syft/src/syft/service/action/action_graph.py [hunk 4]
```diff
@@ -148,7 +149,7 @@ def get(self, uid: Any) -> Any:
     def delete(self, uid: Any) -> None:
         raise NotImplementedError
 
-    def find_neighbors(self, uid: Any) -> List[Any]:
+    def find_neighbors(self, uid: Any) -> Optional[List[Any]]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_graph.py]: should this be
```suggestion
    def find_neighbors(self, uid: UID) -> Optional[List]:
```
? I saw another function with this signature on line 270.

also slight nitpick: could just be `Optional[List]` instead of `Optional[List[Any]]`.

File: packages/syft/tests/syft/request/request_multiple_nodes_test.py (hunk 1 above)
Comment (khoaguin) [reply packages/syft/tests/syft/request/request_multiple_nodes_test.py]: Thanks sure @kiendang

File: packages/syft/src/syft/service/action/action_graph.py (hunk 2 above)
Comment (khoaguin) [reply packages/syft/src/syft/service/action/action_graph.py]: Thanks @kiendang. Done!

File: packages/syft/src/syft/service/action/action_object.py [hunk 5]
```diff
@@ -196,7 +196,9 @@ def syft_history_hash(self) -> int:
         return hashes
 
     @classmethod
-    def syft_function_action_from_kwargs_and_id(cls, kwargs, user_code_id):
+    def syft_function_action_from_kwargs_and_id(
+        cls, kwargs: Any, user_code_id: UID
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_object.py]: should be

```suggestion
        cls, kwargs: Dict[str, Any], user_code_id: UID
```

`**kwargs: Any` but `kwargs: Dict[str, Any]`

File: packages/syft/src/syft/service/action/action_object.py [hunk 6]
```diff
@@ -1260,7 +1266,7 @@ def obj_not_ready(
 
     @staticmethod
     def empty(
-        syft_internal_type: Type[Any] = NoneType,
+        syft_internal_type: Optional[Type[Any]] = NoneType,  # type: ignore[assignment]
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_object.py]:
```suggestion
        syft_internal_type: Optional[Type] = NoneType,  # type: ignore[assignment]
```

File: packages/syft/src/syft/service/action/action_permissions.py [hunk 7]
```diff
@@ -46,7 +46,10 @@ def permission_string(self) -> str:
         if self.permission in COMPOUND_ACTION_PERMISSION:
             return f"{self.permission.name}"
         else:
-            return f"{self.credentials.verify}_{self.permission.name}"
+            if self.credentials is not None:
+                return f"{self.credentials.verify}_{self.permission.name}"
+            else:
+                return f"{self.permission.name}"
```
Comment (kiendang) [inline packages/syft/src/syft/service/action/action_permissions.py]: this can be unnested

```py
if self.credentials is not None:
    return f"{self.credentials.verify}_{self.permission.name}"
return f"{self.permission.name}"
```

File: packages/syft/src/syft/service/code/user_code.py [hunk 8]
```diff
@@ -709,10 +733,10 @@ class SubmitUserCode(SyftObject):
     __repr_attrs__ = ["func_name", "code"]
 
     @property
-    def kwargs(self) -> List[str]:
+    def kwargs(self) -> Optional[dict[Any, Any]]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/code/user_code.py:785]: check if this could be `Optional[dict[str, Any]]`

Comment (khoaguin) [reply packages/syft/src/syft/service/code/user_code.py:785]: I don't think so @kiendang since in the code there are already multiple places that it's defined as `Optional[Dict[Any, Any]]`
Also in line 410 of `user_code.py`, we use it with `[x.verify_key for x in self.input_policy_init_kwargs.keys()]` so it's probably not a string

Comment (kiendang) [reply packages/syft/src/syft/service/code/user_code.py:785]: ah I see

File: packages/syft/src/syft/service/action/action_service.py [hunk 9]
```diff
@@ -291,6 +300,10 @@ def _user_code_execute(
 
         if not override_execution_permission:
             input_policy = code_item.input_policy
+            if input_policy is None:
```
Comment (khoaguin) [inline packages/syft/src/syft/service/action/action_service.py:303]: I added an error handling here to make the `test_code_accept_deny` passed. I think it looks right, right? @koenvanderveen

File: packages/syft/src/syft/util/telemetry.py [hunk 10]
```diff
@@ -85,4 +84,4 @@ def noop(func: T) -> T:
         return noop
 
 
-instrument = setup_tracer()
+instrument: Callable = setup_tracer()
```
Comment (kiendang) [inline packages/syft/src/syft/util/telemetry.py]: This one is tricky. `instrument` must have type `Callable[Concatenate[T, ..], T]`, not just `Callable`, otherwise we lose typing and auto-completion for any class that use the `@instrument` decorator. However, mypy does not seem to be able to infer the type for `instrument` if we just do `instrument = setup_tracer()` and if we set `instrument: Callable[Concatenate[T, ..], T] = setup_tracer()` it complains about wrong typing too. We'd have to modify the code a bit. I'll push a fix for this.

File: packages/syft/src/syft/service/action/action_object.py [hunk 11]
```diff
@@ -659,18 +662,12 @@ def reload_cache(self):
                 blob_retrieval_object = blob_storage_read_method(
                     uid=self.syft_blob_storage_entry_id
                 )
-                if isinstance(blob_retrieval_object, SyftError):
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_object.py:662]: Can we add this back ? and change the return typing for reload_cache to Optional[SyftErrror]

File: packages/syft/src/syft/service/action/action_object.py [hunk 12]
```diff
@@ -775,25 +775,26 @@ def __check_action_data(cls, values: dict) -> dict:
             if inspect.isclass(v):
                 values["syft_action_data_repr_"] = repr_cls(v)
             else:
-                values["syft_action_data_repr_"] = (
-                    v._repr_markdown_()
-                    if hasattr(v, "_repr_markdown_")
-                    else v.__repr__()
-                )
+                if v is not None:
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_object.py]: Any reason for checking v is not `None`. Even if v is None, then also v.__repr__() is a valid argument

File: packages/syft/src/syft/service/action/action_object.py [hunk 13]
```diff
@@ -1710,11 +1719,12 @@ def _repr_markdown_(self) -> str:
             if inspect.isclass(self.syft_action_data_cache):
                 data_repr_ = repr_cls(self.syft_action_data_cache)
             else:
-                data_repr_ = (
-                    self.syft_action_data_cache._repr_markdown_()
-                    if hasattr(self.syft_action_data_cache, "_repr_markdown_")
-                    else self.syft_action_data_cache.__repr__()
-                )
+                if self.syft_action_data_cache is not None:
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_object.py]: Why do we need to check for .syft_action_data_cache to be None ?

File: packages/syft/src/syft/service/action/action_service.py [hunk 14]
```diff
@@ -57,6 +60,8 @@ def __init__(self, store: ActionStore) -> None:
     def np_array(self, context: AuthedServiceContext, data: Any) -> Any:
         if not isinstance(data, np.ndarray):
             data = np.array(data)
+        if context.node is None:
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_service.py]: can we cast here ?

File: packages/syft/src/syft/service/action/action_service.py [hunk 15]
```diff
@@ -291,6 +300,10 @@ def _user_code_execute(
 
         if not override_execution_permission:
             input_policy = code_item.input_policy
+            if input_policy is None:
+                if not code_item.output_policy_approved:
+                    return Err("Execution denied: Your code is waiting for approval")
+                return Err("Your code's input policy is None")
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_service.py]: Can we change the message to say `No input poliicy defined for user code: {code_item.id}`

File: packages/syft/src/syft/service/code/user_code.py [hunk 16]
```diff
@@ -425,22 +431,35 @@ def is_enclave_code(self) -> bool:
...
+            return [str(x.node_name) for x in self.input_policy_init_kwargs.keys()]
+        return None
 
     @property
-    def input_owner_verify_keys(self) -> List[SyftVerifyKey]:
-        return [x.verify_key for x in self.input_policy_init_kwargs.keys()]
+    def input_owner_verify_keys(self) -> Optional[List[SyftVerifyKey]]:
+        if self.input_policy_init_kwargs is not None:
+            return [x.verify_key for x in self.input_policy_init_kwargs.keys()]
+        return None
```
Comment (shubham3121) [inline packages/syft/src/syft/service/code/user_code.py:445]: we can skip returning None, its the default behaviour with no return types.

File: packages/syft/src/syft/service/code/user_code.py [hunk 17]
```diff
@@ -571,11 +592,13 @@ def assets(self) -> List[Asset]:
         if api is None:
             return SyftError(message=f"You must login to {self.node_uid}")
 
-        inputs = (
-            uids
-            for node_identity, uids in self.input_policy_init_kwargs.items()
-            if node_identity.node_name == api.node_name
-        )
+        if self.input_policy_init_kwargs is not None:
```
Comment (shubham3121) [inline packages/syft/src/syft/service/code/user_code.py:598]: We're missing defining inputs if `self.input_policy_init_kwargs` is None

```python


inputs = () 
if self.input_policy_init_kwargs is not None:
    inputs = (
                uids
                for node_identity, uids in self.input_policy_init_kwargs.items()
                if node_identity.node_name == api.node_name
            )

```

File: packages/syft/src/syft/service/code/user_code.py [hunk 18]
```diff
@@ -824,39 +852,40 @@ def _ephemeral_node_call(
...
-                new_obj = ActionObject.from_obj(
-                    data_obj.syft_action_data,
-                    id=id,
-                    syft_node_location=node_id.node_id,
-                    syft_client_verify_key=node_id.verify_key,
-                )
-                res = ep_client.api.services.action.set(new_obj)
-                if isinstance(res, SyftError):
-                    return res
+        if self.input_policy_init_kwargs is not None:
```
Comment (shubham3121) [inline packages/syft/src/syft/service/code/user_code.py]: can we type cast here `cast(self.input_policy_init_kwargs, dict)`

**Review by shubham3121 — CHANGES_REQUESTED** [review]

File: packages/syft/src/syft/service/action/action_object.py (hunk 12 above)
Comment (shubham3121) [reply packages/syft/src/syft/service/action/action_object.py]: Fixed by: 4961d8bd46eb13279bb7112cb5615f5f9ec8b6a5

File: packages/syft/src/syft/service/action/action_object.py [hunk 19]
```diff
@@ -914,7 +915,7 @@ def _syft_try_to_save_to_store(self, obj) -> None:
         )
 
         if api is not None:
-            TraceResult.result += [action]
+            TraceResult.result += [action]  # type: ignore[unreachable]
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_object.py]: fixed by: dd52fb249c90f35055c440ddd403dff512c6d848

File: packages/syft/src/syft/service/action/action_object.py [hunk 20]
```diff
@@ -894,10 +895,10 @@ def _syft_try_to_save_to_store(self, obj) -> None:
 
         api = None
         if TraceResult._client is not None:
-            api = TraceResult._client.api
+            api = TraceResult._client.api  # type: ignore[unreachable]
 
         if api is not None:
-            obj._set_obj_location_(api.node_uid, api.signing_key.verify_key)
+            obj._set_obj_location_(api.node_uid, api.signing_key.verify_key)  # type: ignore[unreachable]
```
Comment (shubham3121) [inline packages/syft/src/syft/service/action/action_object.py]: fixed by: dd52fb249c90f35055c440ddd403dff512c6d848

File: packages/syft/src/syft/service/action/action_object.py (hunk 13 above)
Comment (shubham3121) [reply packages/syft/src/syft/service/action/action_object.py]: fixed by: 1e952fd0ef4e0f4f5be494ce530e7c2adacac7b9

**Review by shubham3121 — APPROVED** [review]
@khoaguin this is tremendous work !! 🦾 💯
